Fix / Enhancement for Omitted Object Literals

torch2424 opened the pull request:
closes #1226

This fixes the bug where omitted Object literals would just give random memory from the runtime. By introducing an error for class fields that cannot be omitted.

This Enhances Object literals, by allowing the [basic AS Types](https://docs.assemblyscript.org/basics/types). To be omitted, with default values of zero :smile: 

TODO: Either we add support for nullable types to default to null and/or add support for making a new instance of a class if the constructor has no parameters in this same PR. Or we can do that work in a seperate PR :smile: 

cc @dcodeIO @jtenner 

P.S. Yay! My first real AS Compiler PR :smile: :tada: 

# Screenshots

![Screenshot from 2020-04-16 17-45-32](https://user-images.githubusercontent.com/1448289/79520144-d59bb080-800a-11ea-88bb-e321a30efd5d.png)
![Screenshot from 2020-04-16 17-45-06](https://user-images.githubusercontent.com/1448289/79520146-d6344700-800a-11ea-97a4-a1ff7747ad25.png)
![Screenshot from 2020-04-16 17-44-29](https://user-images.githubusercontent.com/1448289/79520155-d9c7ce00-800a-11ea-8a63-9b3dc1d22a6b.png)

torch2424 commented:
Oops, seems like I am failing the self-compilation tests :upside_down_face:  

@dcodeIO did you get things self compiling? Should I go back and fix these? Shouldn't be too hard, all I think I really need to do is get rid of the phat arraw function I added :smile: But how do I test the self compiler from my machine? :thinking:

dcodeIO commented on src/compiler.ts:
```diff
@@ -8682,12 +8682,36 @@ export class Compiler extends DiagnosticEmitter {
+    // Assume all class fields will be omitted, and add them to our omitted list
+    var omittedClassFieldMembers = new Map();
+    if (members) {
+      for (let memberKey of members.keys()) {
```
```suggestion
      for (let _keys = Map_keys(members), i = 0, k = _keys.length; i < k; ++i) {
        let memberKey = _keys[i];
```
Necessary because we don't have iterators (and thus no `for ... of`) yet.

dcodeIO commented on src/compiler.ts:
```diff
@@ -8682,12 +8682,36 @@ export class Compiler extends DiagnosticEmitter {
+      for (let memberKey of members.keys()) {
+        let member = members.get(memberKey);
+        if (member && member.kind == ElementKind.FIELD) {
+          omittedClassFieldMembers.set(member.name, '');
```
Wondering if this can be a `Set`, if initial values are empty?

torch2424 replied:
Yeah I originally was thinking a set, but then I got distracted and used a map when I came back :upside_down_face: Will switch to a set! :smile:

torch2424 replied:
Oh wait! I remember, I chose not to use a set, because Set.values() returns am Iterator in JS, but an Array in AS :upside_down_face:

dcodeIO replied:
Yeah, `Map`s and `Set`s have the same problem there. For `Set`s, the helper is `Set_values`. Both can't be iterated in a uniform way in TS/AS yet (no `for ... of`).

torch2424 replied:
Oh wait ignore me, Map.keys() also returns an iterator haha!

dcodeIO commented on src/compiler.ts:
```diff
@@ -8682,12 +8682,36 @@ export class Compiler extends DiagnosticEmitter {
+    if (members) {
+      for (let memberKey of members.keys()) {
+        let member = members.get(memberKey);
+        if (member && member.kind == ElementKind.FIELD) {
```
```suggestion
        if (member !== null && member.kind == ElementKind.FIELD) {
```
That's necessary due to some obscurities when compiling logical expressions. This way, all parts are `bool`.

dcodeIO commented on src/compiler.ts:
```diff
@@ -8682,12 +8682,36 @@ export class Compiler extends DiagnosticEmitter {
+    }
+
+    // Define a function to add a new expression on our Object literal
+    let storeObjectLiteralField = (fieldInstance: Field, fieldType: Type, valueExpr: ExpressionRef) => {
+      exprs.push(this.module.store( // TODO: handle setters as well
+        fieldType.byteSize,
+        this.module.local_get(tempLocal.index, this.options.nativeSizeType),
+        valueExpr,
+        fieldType.toNativeType(),
+        fieldInstance.memoryOffset
+      ));
+    };
```
Suggesting to inline this to match other code. Also, having this in-line avoids an indirect call.

dcodeIO commented on src/compiler.ts:
```diff
@@ -8700,27 +8724,114 @@ export class Compiler extends DiagnosticEmitter {
 
+    // Iterate through the remaining omittedClassFieldMembers.
+    if (members) {
+      for(let omittedClassFieldMemberKey of omittedClassFieldMembers.keys()) {
```
```suggestion
      for (let _keys = Map_keys(omittedClassFieldMemberKey), i = 0, k = _keys.length; i < k; ++i) {
        let omittedClassFieldMemberKey = _keys[i];
```

dcodeIO commented on src/compiler.ts:
```diff
@@ -8700,27 +8724,114 @@ export class Compiler extends DiagnosticEmitter {
+    // Iterate through the remaining omittedClassFieldMembers.
+    if (members) {
+      for(let omittedClassFieldMemberKey of omittedClassFieldMembers.keys()) {
+        let member = members.get(omittedClassFieldMemberKey);
```
```suggestion
        let member = assert(members.get(omittedClassFieldMemberKey));
```
This is mostly for TypeScript, where the result may be `undefined`. Since the key is in `omittedClassFields` here, we know that this will never be the case.

dcodeIO commented on src/compiler.ts:
```diff
@@ -8700,27 +8724,114 @@ export class Compiler extends DiagnosticEmitter {
+      for(let omittedClassFieldMemberKey of omittedClassFieldMembers.keys()) {
+        let member = members.get(omittedClassFieldMemberKey);
+        
+        if(member) {
```
```suggestion
        if (member) {
```

dcodeIO commented on src/compiler.ts:
```diff
@@ -8700,27 +8724,114 @@ export class Compiler extends DiagnosticEmitter {
+          let fieldInstance = <Field>member;
+          let fieldType = fieldInstance.type;
+
+          if (fieldType.classReference) {
```
```suggestion
          if (fieldType.is(TypeFlags.REFERENCE) && fieldType.classReference !== null) {
```
There may be types that have a `classReference` (at some point), but are not actually classes, like the wrapper class `I32` of `i32`.

dcodeIO commented on src/compiler.ts:
```diff
@@ -8700,27 +8724,114 @@ export class Compiler extends DiagnosticEmitter {
+
+            // Otherwise, error
+            this.error(
+              DiagnosticCode.Object_literal_is_missing_class_member_fields_that_must_be_defined,
```
This looks like a custom error message. What I typically do is look through https://github.com/microsoft/TypeScript/blob/master/src/compiler/diagnosticMessages.json and check if there is an existing one we can reuse, if we don't have it already. Only if there is none we can reuse, I add a new one.

torch2424 replied:
@dcodeIO I looked through the errors (CTRL+F for "literal") and I couldn't find anything that matched. :smile: 

I feel like this error will be needed, even after we add support for nullablity and things, since I expect some classes will have funny stuff with their constructors and things? :thinking: I purposely tried to make it vague enough to be like, Hey, your missing a field that you need to add manually haha!

dcodeIO replied:
Perhaps, if there is a field on the object that is not on the class type

> `"Property '{0}' does not exist on type '{1}'.": 2339`

and if there is a missing field

> `"Property '{0}' is missing in type '{1}' but required in type '{2}'.": 2741`

with `{1}` being `<object>` and `{2}` being the class type?

torch2424 replied:
Dope! Yeah that sounds good, I will use those error messages! :smile:

dcodeIO commented:
Regarding self-compilation: Yes, this is indeed part of CI now. The relevant scripts are `npm run asbuild` to build it, and `npm run astest` to test it. Unless on top of the vtable branch, the test will not do much and ultimately fail, which is expected, while on top of the vtable branch, it is good enough to output a simple module in text format.

MaxGraey commented:
Isn't this relate to #1155 ?

torch2424 commented:
@dcodeIO Thanks for the quick review! :smile: :+1:  Gonna start addressing comments in a little bit!

@MaxGraey Ummmm, so I am not too sure? I think it's similar, but a different issue? I'll let @dcodeIO answer that.

dcodeIO commented:
It's somewhat related conceptually, yeah. Let's say we call the (default) ctor on object creation, then the ctor may initialize some of these fields, so specifying these is not necessary even though the field is not nullable. However, we are not calling the ctor here even though we should.

Suggesting to do the following to be safe for now: As soon as the class has a constructor, disallow creating it from an object literal. This should avoid any issues, and we can look into actually calling ctors in a later PR.

jtenner commented:
It's unfortunate that we aren't like rust, and get to have traits where object literals might make sense.

torch2424 commented:
@dcodeIO Made all requested changes, and asbuild works on my machine :smile: :+1: This is good to go!

![Screenshot from 2020-04-17 15-53-45](https://user-images.githubusercontent.com/1448289/79620119-bdd43300-80c3-11ea-8ef6-9d665d7734b2.png)

dcodeIO commented on src/compiler.ts:
```diff
@@ -8682,12 +8682,26 @@ export class Compiler extends DiagnosticEmitter {
     assert(numNames == values.length);
+
+    // Assume all class fields will be omitted, and add them to our omitted list
+    let omittedClassFieldMembers = new Map<string, string>();
```
```suggestion
    let omittedClassFieldMembers = new Set<string>();
```
Let's make this a `Set` and use `_values = Set_values` instead of `_keys = Map_keys` below.

dcodeIO commented on src/compiler.ts:
```diff
@@ -8700,27 +8714,122 @@ export class Compiler extends DiagnosticEmitter {
-        fieldType.toNativeType(),
-        fieldInstance.memoryOffset
-      );
+      exprs.push(this.module.store( // TODO: handle setters as well
+       fieldType.byteSize,
+       this.module.local_get(tempLocal.index, this.options.nativeSizeType),
+       this.compileExpression(values[i], fieldInstance.type, Constraints.CONV_IMPLICIT),
+       fieldType.toNativeType(),
+       fieldInstance.memoryOffset
+      ));
```
This looks problematic, but already was so before so nothing introduced by this PR. If `fieldType` is managed here, and the class is managed as well, we need to retain a reference that the instance now holds, ideally accounting for pre-retained values (from function calls or ctors) which we can trigger by `WILL_RETAIN` and checking `skippedAutoreleases`.

```suggestion
      let expr = this.compileExpression(values[i], fieldType, Constraints.CONV_IMPLICIT | Constraints.WILL_RETAIN);
      if (isManaged && fieldType.isManaged && !this.skippedAutoreleases.has(expr)) {
        expr = this.makeRetain(expr);
      }
      exprs.push(this.module.store( // TODO: handle setters as well
        fieldType.byteSize,
        this.module.local_get(tempLocal.index, this.options.nativeSizeType),
        expr,
        fieldType.toNativeType(),
        fieldInstance.memoryOffset
      ));
```

dcodeIO commented on src/compiler.ts:
```diff
@@ -8700,27 +8714,122 @@ export class Compiler extends DiagnosticEmitter {
+
+          // Otherwise, error
+          this.error(
+            DiagnosticCode.Object_literal_is_missing_class_member_fields_that_must_be_defined,
```
I made a suggesting for reusable error messages in a comment below. What do you think of it?

dcodeIO commented on src/compiler.ts:
```diff
@@ -8700,27 +8714,122 @@ export class Compiler extends DiagnosticEmitter {
+            exprs.push(this.module.store( // TODO: handle setters as well
+             fieldType.byteSize,
+             this.module.local_get(tempLocal.index, this.options.nativeSizeType),
+             this.module.i32(0),
```
Most of this can be simplified by using 

```suggestion
             this.makeZero(fieldType),
```

so we don't need all these cases.

jtenner commented:
Nice work @torch2424

torch2424 commented:
Thanks @jtenner ! :smile: :tada: 

And made the requested changes @dcodeIO ! :smile: 

![Screenshot from 2020-04-20 11-31-32](https://user-images.githubusercontent.com/1448289/79786690-b9de2600-82fa-11ea-8d6e-47678390955a.png)
![Screenshot from 2020-04-20 11-31-15](https://user-images.githubusercontent.com/1448289/79786694-ba76bc80-82fa-11ea-864a-b3468a46520c.png)
![Screenshot from 2020-04-20 11-30-07](https://user-images.githubusercontent.com/1448289/79786696-bb0f5300-82fa-11ea-9c07-bda74a9a0bf3.png)

dcodeIO commented on src/compiler.ts:
```diff
@@ -8700,27 +8714,94 @@ export class Compiler extends DiagnosticEmitter {
+        expr = this.makeRetain(expr);
+      }
+      exprs.push(this.module.store( // TODO: handle setters as well
+       fieldType.byteSize,
+       this.module.local_get(tempLocal.index, this.options.nativeSizeType),
+       expr,
+       fieldType.toNativeType(),
+       fieldInstance.memoryOffset
```
nit: indentation

dcodeIO commented on src/compiler.ts:
```diff
@@ -8700,27 +8714,94 @@ export class Compiler extends DiagnosticEmitter {
 
+    // Iterate through the remaining omittedClassFieldMembers.
+    if (members) {
+      for(let _values = Set_values(omittedClassFieldMembers), i = 0, v = _values.length; i < v; ++i) {
+        let omittedMemberKey = _values[i];
```
nit: indentation (and other code tends to use j/l)
```suggestion
      for (let _values = Set_values(omittedClassFieldMembers), j = 0, l = _values.length; j < l; ++j) {
        let omittedMemberKey = _values[j];
```

dcodeIO commented on src/compiler.ts:
```diff
@@ -8700,27 +8714,94 @@ export class Compiler extends DiagnosticEmitter {
+
+        if (fieldType.is(TypeFlags.REFERENCE) && fieldType.classReference !== null) {
+          // TODO: Check if it is a class, with a default value (constructor with no params).
+          // TODO: Check if it can be null, and set to null
+
+          // Otherwise, error
+          this.error(
+            DiagnosticCode.Property_0_is_missing_in_type_1_but_required_in_type_2,
+            expression.range, "<object>", classReference.toString()
+          );
+          hasErrors = true;
+          continue;
```
Reminds me that you asked for this for your use case, and this can be supported by only erroring if the field's type is not nullable, and otherwise letting it store a `0` below:

```suggestion
          if (!fieldType.is(TypeFlags.NULLABLE)) {
            this.error(
              DiagnosticCode.Property_0_is_missing_in_type_1_but_required_in_type_2,
              expression.range, "<object>", classReference.toString()
            );
            hasErrors = true;
            continue;
          }
```

torch2424 replied:
Oooooo nice! Thank you for suggesting that quick fix for support! :smile: :+1:

dcodeIO commented on src/compiler.ts:
```diff
@@ -8700,27 +8714,94 @@ export class Compiler extends DiagnosticEmitter {
+            ));
+            continue;
+          }
+          default: {}
```
nit: This line can be removed.

dcodeIO commented on src/compiler.ts:
```diff
@@ -8700,27 +8714,94 @@ export class Compiler extends DiagnosticEmitter {
+          case TypeKind.F32: 
+          case TypeKind.F64: {
+            exprs.push(this.module.store( // TODO: handle setters as well
+             fieldType.byteSize,
+             this.module.local_get(tempLocal.index, this.options.nativeSizeType),
+             this.makeZero(fieldType),
+             fieldType.toNativeType(),
+             fieldInstance.memoryOffset
```
nit: indentation

torch2424 replied:
My Vim was not very friendly to me :joy:

dcodeIO commented on src/compiler.ts:
```diff
@@ -8682,12 +8682,26 @@ export class Compiler extends DiagnosticEmitter {
+      }
+    }
+
+    // Iterate through the members definted in our expression
```
nit: typo in "defined"

torch2424 commented:
@dcodeIO Made requested changes, tests still pass, asbuild still working :smile: :+1:

dcodeIO commented:
Thank you! :)

torch2424 commented:
Yayyyy! :smile: @dcodeIO Thank you very much for the review. Literally tweeting this out now :joy:
